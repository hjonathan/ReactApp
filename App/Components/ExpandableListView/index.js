import React, {Component} from "react";
import {Text, View, Image, TouchableHighlight, Animated} from "react-native";
import * as Res from "../../Assets/resources";
import Styles from "./styles";

class ListView extends Component {
    constructor (props) {
        super(props);

        this.icons = {
            up: Res.images.arrowHeadUp,
            down: Res.images.arrowHeadDown
        };

        this.state = {
            title: props.title,
            expanded: true
        };
        this.toggle = this.toggle.bind(this);
    }

    /**
     * Change status expanded for Content
     */
    toggle () {
        this.setState({
            expanded: !this.state.expanded
        });
    }

    /**
     * Renders
     * @returns {*}
     */
    render () {
        let icon = this.icons.down;

        if (this.state.expanded) {
            icon = this.icons.up;
        }

        return (
            <Animated.View style={Styles.container}>
                <View style={Styles.titleContainer}>
                    <Text style={Styles.title}>{this.state.title}</Text>
                    <TouchableHighlight
                        style={Styles.button}
                        onPress={this.toggle}
                        underlayColor="#f1f1f1"
                    >
                        <Image
                            style={Styles.buttonImage}
                            source={icon}
                        />
                    </TouchableHighlight>
                </View>
                {
                    this.state.expanded && (
                        <View style={Styles.body}>
                            {this.props.children}
                        </View>
                    )
                }
            </Animated.View>
        );
    }
}

export default ListView;
