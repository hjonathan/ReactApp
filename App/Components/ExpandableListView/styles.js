import {StyleSheet} from "react-native";

export default StyleSheet.create({
    container: {
        overflow: "hidden",
        paddingBottom: 10
    },
    titleContainer: {
        backgroundColor: "#fff",
        flexDirection: "row"
    },
    title: {
        flex: 1,
        padding: 10,
        color: "#3397E1"
    },
    button: {
        justifyContent: "center"
    },
    buttonImage: {
        width: 30,
        height: 25
    },
    body: {
        padding: 10,
        paddingTop: 0
    }
});
