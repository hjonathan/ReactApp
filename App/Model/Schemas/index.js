import Realm from "realm";
import CounterSchema from "./Counter";
import * as Lists from "./List";
import TodoSchema from "./Todo";
import DraftSchema from "./Draft";
import ParticipatedSchema from "./Participated";
import UnassignedSchema from "./Unassigned";
import StartCaseSchema, {StartCaseStepSchema} from "./StartCase";
import UserSchema from "./User";
import StepSchema from "./Step";
import StepTriggerSchema from "./StepTrigger";
import FormSchema from "./Form";
import NewCaseSchema from "./NewCase";
import DataSchema from "./Data";
import MobileControlSchema from "./MobileControl";

/**
 * Return the Realm instance
 * @type {Realm}
 */
export const RealmObject = new Realm({
    schema: [
        UserSchema,
        CounterSchema,
        Lists.ProcessSchema,
        Lists.TaskSchema,
        Lists.UserOwnerSchema,
        Lists.UsersSchema,
        TodoSchema,
        DraftSchema,
        ParticipatedSchema,
        UnassignedSchema,
        StartCaseSchema,
        StartCaseStepSchema,
        StepSchema,
        StepTriggerSchema,
        FormSchema,
        NewCaseSchema,
        DataSchema,
        MobileControlSchema
    ],
    schemaVersion: 1.0
});

/**
 * Return all Schemas
 */
export {
    UserSchema,
    CounterSchema,
    TodoSchema,
    DraftSchema,
    ParticipatedSchema,
    UnassignedSchema,
    StartCaseSchema,
    StepSchema,
    FormSchema,
    NewCaseSchema,
    DataSchema,
    MobileControlSchema
};
