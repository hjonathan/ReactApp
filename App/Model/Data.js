import _ from "lodash";
import moment from "moment";
import BaseModel from "./Interface/BaseModel";
import {RealmObject, DataSchema} from "./Schemas";
import {CASE, DATETIME} from "../Libs/Const";

/**
 * Model for Data
 */
class Data extends BaseModel {
    constructor () {
        super(DataSchema.schema.name);
    }

    /**
     * Save the list when the action formData/server/update is executed
     * @param params
     */
    saveData (params) {
        let saved = null,
            date = moment().format(DATETIME.LOCAL),
            data;
        RealmObject.write(() => {
            let oldData = {},
                dbCaseData = RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${params.app_uid}"`);
            if (dbCaseData.length > 0) {
                oldData = _.values(dbCaseData)[0].dataJSON;
            }
            data = {
                caseId: params.app_uid,
                caseIndex: Number(params.del_index),
                formId: params.dyn_uid,
                data: JSON.stringify(Object.assign({}, oldData, params.data)),
                createDate: moment(params.createDate).format(DATETIME.LOCAL) || date,
                updateDate: date,
                status: CASE.STATUS.OFFLINE
            };
            saved = RealmObject.create(DataSchema.schema.name, data, true);
        });
        return {saved};
    }

    /**
     * Save case data when the action cases/data/request is executed
     * @param payload
     * @param state
     * @returns {{saved: *}}
     */
    saveCaseData (payload, state) {
        const {params, response} = payload;
        let saved = null,
            date = moment().format(DATETIME.LOCAL),
            data;
        RealmObject.write(() => {
            let oldData = {},
                dbCaseData = RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${params.app_uid}"`);
            if (dbCaseData.length > 0) {
                oldData = _.values(dbCaseData)[0].dataJSON;
            }
            data = {
                caseId: params.app_uid,
                caseIndex: Number(params.del_index),
                data: JSON.stringify(Object.assign({}, oldData, response)),
                formId: params.formId || "",
                stepId: params.stepId || "",
                stepHistory: JSON.stringify(params.stepHistory) || "[]",
                createDate: moment(params.createDate).format(DATETIME.LOCAL) || date,
                updateDate: date,
                type: params.type || state.Cases.routeName,
                status: params.status || CASE.STATUS.NONE
            };
            saved = RealmObject.create(DataSchema.schema.name, data, true);
        });
        return {saved};
    }

    /**
     * Get data
     * @param caseId
     * @returns {*}
     */
    getDataCase (caseId) {
        return _.values(RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${caseId}"`)).shift();
    }

    /**
     * Get all data in status offline
     * @param caseId
     * @returns {*}
     */
    getAllCaseOffline () {
        return _.values(RealmObject.objects(DataSchema.schema.name).filtered(`status = "${CASE.STATUS.OFFLINE}"`));
    }

    /**
     * Get data of case with caseId and status is equal to working (work in progress)
     * @param caseId
     * @returns {*}
     */
    filterCaseIdWIP (caseId) {
        const result = RealmObject.objects(DataSchema.schema.name)
            .filtered(`caseId = "${caseId}" AND status = "${CASE.STATUS.WORKING}"`);
        return _.values(result.map((x) => {
            let newObject = _.assign({}, x);
            newObject.stepHistory = x.stepHistoryArray;
            newObject.data = x.dataJSON;
            return newObject;
        })).shift();
    }

    /**
     * Change status of row
     * @param caseId
     * @param status
     */
    changeStatus (caseId, status) {
        let dbData;
        RealmObject.write(() => {
            dbData = _.values(RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${caseId}"`)).shift();
            if (dbData) {
                dbData.status = status;
                RealmObject.create(DataSchema.schema.name, dbData, true);
            }
        });
    }

    /**
     * Remove the specified data case of storage.
     * @param value
     */
    destroyDataCase (value) {
        RealmObject.write(() => {
            let data = RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${value}"`);
            RealmObject.delete(data);
        });
    }
    /**
     * Delete multiple Form
     */
    destroyObject () {
        RealmObject.write(() => {
            let allObjects = RealmObject.objects(DataSchema.schema.name);
            RealmObject.delete(allObjects); // Deletes all Forms
        });
    }
}

export default new Data();
