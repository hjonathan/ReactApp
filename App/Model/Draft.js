import _ from "lodash";
import BaseModel from "./Interface/BaseModel";
import {RealmObject, DraftSchema, DataSchema} from "./Schemas";
import {CASE} from "../Libs/Const";

/**
 * Model for list Draft
 */
class DraftCase extends BaseModel {
    constructor () {
        super(DraftSchema.schema.name);
    }

    /**
     * Save the list when the action cases/draft/success is executed
     * @param listDraft
     */
    saveList (listDraft) {
        let dbData;
        RealmObject.write(() => {
            listDraft.data.response.map((draftCase) => {
                let otherCase = RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${draftCase.caseId}" AND type != "${CASE.DRAFT}"`);
                RealmObject.delete(otherCase);
                dbData = _.values(RealmObject.objects(DataSchema.schema.name).filtered(`caseId = "${draftCase.caseId}" AND type = "${CASE.DRAFT}"`)).shift();
                if (dbData) {
                    draftCase.status = dbData.status;
                    RealmObject.create(DraftSchema.schema.name, draftCase, true);
                    return;
                }
                RealmObject.create(DraftSchema.schema.name, draftCase, true);
                return;
            });
        });
    }

    /**
     * Return all rows of list Draft
     */
    getDraftOffline () {
        let result = RealmObject.objects(DraftSchema.schema.name).sorted("delegateDate", true);
        return result.map(x => _.assign({}, x));
    }

    /**
     * Return all rows in status sending
     * @returns {any[]}
     */
    getDraftSending () {
        let result = RealmObject.objects(DraftSchema.schema.name).filtered(`status = "${CASE.STATUS.SENDING}"`);
        return result.map(x => _.assign({}, x));
    }

    /**
     * Return a rows of case Draft
     */
    getDraftCase (caseId) {
        return _.values(RealmObject.objects(DraftSchema.schema.name).filtered(`caseId = "${caseId}"`)).shift();
    }

    /**
     * Check if a cases exist in Draft
     * @param caseId
     * @returns {boolean}
     */
    existCase (caseId) {
        const result = RealmObject.objects(DraftSchema.schema.name).filtered(`caseId = "${caseId}"`);
        return result.length > 0;
    }

    /**
     * Change status of row
     * @param caseId
     * @param status
     */
    changeStatus (caseId, status) {
        let dbDraft;
        RealmObject.write(() => {
            dbDraft = _.values(RealmObject.objects(DraftSchema.schema.name).filtered(`caseId = "${caseId}"`)).shift();
            if (dbDraft) {
                dbDraft.status = status;
                RealmObject.create(DraftSchema.schema.name, dbDraft, true);
            }
        });
    }

    /**
     * Remove the specified data case of storage.
     * @param value
     */
    destroyCase (value) {
        RealmObject.write(() => {
            let data = RealmObject.objects(DraftSchema.schema.name).filtered(`caseId = "${value}"`);
            RealmObject.delete(data);
        });
    }

    /**
     * Delete multiple Case in Draft
     */
    destroyObject () {
        RealmObject.write(() => {
            let allObjects = RealmObject.objects(DraftSchema.schema.name);
            RealmObject.delete(allObjects); // Deletes all case in Draft
        });
    }
}

export default new DraftCase();
