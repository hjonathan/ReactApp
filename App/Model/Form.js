import _ from "lodash";
import BaseModel from "./Interface/BaseModel";
import {RealmObject, FormSchema} from "./Schemas";

/**
 * Model for Form
 */
class FormCase extends BaseModel {
    constructor () {
        super(FormSchema.schema.name);
    }

    /**
     * Save the list when the action dynaform/processed/success is executed
     * @param forms
     */
    saveList (forms) {
        RealmObject.write(() => {
            let form = forms.status.response;
            if (!forms.status.params.app_uid && !forms.status.params.del_index) {
                RealmObject.create(FormSchema.schema.name, Object.assign({}, form, {
                    formContent: JSON.stringify(form.formContent)
                }), true);
            }
        });
    }

    /**
     * Returns all the forms that match the formId
     */
    getFormOffline (formId) {
        let forms = _.values(RealmObject.objects(FormSchema.schema.name).filtered(`formId = "${formId}"`))[0],
            response = {};
        if (forms) {
            response.formId = forms.formId;
            response.formTitle = forms.formTitle;
            response.formDescription = forms.formDescription;
            response.formContent = forms.formContentJSON;
            response.formUpdateDate = forms.formUpdateDate;
        }
        return _.isEmpty(response) ? null : response;
    }

    /**
     * Delete multiple Form
     */
    destroyObject () {
        RealmObject.write(() => {
            let allObjects = RealmObject.objects(FormSchema.schema.name);
            RealmObject.delete(allObjects); // Deletes all Forms
        });
    }
}

export default new FormCase();
