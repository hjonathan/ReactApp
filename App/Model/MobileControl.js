import _ from "lodash";
import BaseModel from "./Interface/BaseModel";
import {RealmObject, MobileControlSchema} from "./Schemas";

/**
 * Mobile control model
 */
class MobileControl extends BaseModel {
    constructor () {
        super(MobileControlSchema.schema.name);
    }

    /**
     * Adds new location
     * @param control
     */
    addLocation (control) {
        RealmObject.write(() => {
            RealmObject.create(MobileControlSchema.schema.name, control, true);
        });
    }

    /**
     * Gets all locations
     * @returns {Array}
     */
    getLocations () {
        return _.values(RealmObject.objects(MobileControlSchema.schema.name));
    }

    /**
     * Filters id by attribute
     * @param idLoc
     * @returns {Array}
     */
    filterBy (idLoc) {
        return _.values(RealmObject.objects(MobileControlSchema.schema.name).filtered(`caseId="${idLoc}"`));
    }

    /**
     * Removes specific location by idField
     * @param idControl
     */
    removeLocation (idControl) {
        let location;
        RealmObject.write(() => {
            location = RealmObject.objects(MobileControlSchema.schema.name).filtered(`idField="${idControl}"`);
            RealmObject.delete(location);
        });
    }

    /**
     * Removes data model
     */
    destroyControl () {
        let locations;
        RealmObject.write(() => {
            locations = RealmObject.objects(MobileControlSchema.schema.name);
            RealmObject.delete(locations);
        });
    }
}

export default new MobileControl();
