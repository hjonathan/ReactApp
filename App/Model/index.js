import Step from "./Step";
import User from "./User";
import StartCase from "./StartCase";
import Todo from "./Todo";
import Unassigned from "./Unassigned";
import Participated from "./Participated";
import Draft from "./Draft";
import Form from "./Form";
import NewCase from "./NewCase";
import Data from "./Data";
import Task from "./Task";
import MobileControl from "./MobileControl";

/**
 * Returns the models
 */
export {
    Step,
    User,
    StartCase,
    Todo,
    Unassigned,
    Participated,
    Draft,
    Form,
    NewCase,
    Data,
    Task,
    MobileControl
};
