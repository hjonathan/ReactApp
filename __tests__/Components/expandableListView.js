import 'react-native';
import React from 'react';
import {RealmObject} from "../../App/Model/Schemas";
import ExpandableListView from '../../App/Components/ExpandableListView';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

describe('Test of Expandable List View card for list', () => {

    let store, container;
    const initialState = {
            title: "Test List"
        },
        mockStore = configureStore();

    beforeEach(() => {
        store = mockStore(initialState);
        container = shallow(<ExpandableListView store={store}/>)
    });

    afterAll(() => {
        RealmObject.close();
    });

    it('render Item correctly', () => {
        expect(toJson(container)).toMatchSnapshot();
    });

    it('Render the connected(SMART) component', () => {
        expect(container.length).toEqual(1)
    });

    it('toggle function', () => {
        expect(container.instance().toggle()).toMatchSnapshot();
    });
})
