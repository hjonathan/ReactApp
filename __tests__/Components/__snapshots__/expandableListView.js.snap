// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of Expandable List View card for list render Item correctly 1`] = `
<AnimatedComponent
  style={
    Object {
      "overflow": "hidden",
      "paddingBottom": 10,
    }
  }
>
  <Component
    style={
      Object {
        "backgroundColor": "#fff",
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#3397E1",
          "flex": 1,
          "padding": 10,
        }
      }
    />
    <TouchableHighlight
      activeOpacity={0.85}
      delayPressOut={100}
      onPress={[Function]}
      style={
        Object {
          "justifyContent": "center",
        }
      }
      underlayColor="#f1f1f1"
    >
      <Image
        source={
          Object {
            "testUri": "../../../App/Assets/icons/arrowHeadUp.png",
          }
        }
        style={
          Object {
            "height": 25,
            "width": 30,
          }
        }
      />
    </TouchableHighlight>
  </Component>
  <Component
    style={
      Object {
        "padding": 10,
        "paddingTop": 0,
      }
    }
  />
</AnimatedComponent>
`;

exports[`Test of Expandable List View card for list toggle function 1`] = `undefined`;
